hbase-server/src/test/java/org/apache/hadoop/hbase/client/TestHMasterAsyncRegistry.java

```diff
@@ -0,0 +1,140 @@
+import static org.junit.Assert.assertEquals;
+
+@Category({ LargeTests.class, ClientTests.class })
+public class TestHMasterAsyncRegistry {
+  private static final Logger LOG = LoggerFactory.getLogger(TestHMasterAsyncRegistry.class);
+  private static final HBaseTestingUtility TEST_UTIL = new HBaseTestingUtility();
+  private static final TableName TEST_TABLE = TableName.valueOf("foo");
+  private static final byte[] COL_FAM = Bytes.toBytes("cf");
+  private static final byte[] QUALIFIER = Bytes.toBytes("col");
+
+  @ClassRule
+  public static final HBaseClassTestRule CLASS_RULE =
+      HBaseClassTestRule.forClass(TestHMasterAsyncRegistry.class);
+
+  // Automatically cleaned up on cluster shutdown.
+  private static AsyncClusterConnection customConnection;
+  private static HMasterAsyncRegistry REGISTRY;
+
+  @BeforeClass
+  public static void setUp() throws Exception {
+    TEST_UTIL.getConfiguration().setInt(META_REPLICAS_NUM, 3);
+    TEST_UTIL.startMiniCluster(3);
+    Configuration conf = new Configuration(TEST_UTIL.getConfiguration());
+    List<HostAndPort> hostAndPorts = new ArrayList<>();
+    String masterHostName =
+        TEST_UTIL.getMiniHBaseCluster().getMaster().getServerName().getHostname();
+    int masterPort = TEST_UTIL.getMiniHBaseCluster().getMaster().getServerName().getPort();
+    // Add some invalid hostAndPort configs. The implementation should be resilient enough to
+    // skip those and probe for the only working master.
+    // 1. Valid hostname but invalid port
+    hostAndPorts.add(HostAndPort.fromParts(masterHostName, 10001));
+    // 2. Invalid hostname but valid port
+    hostAndPorts.add(HostAndPort.fromParts("foo.bar", masterPort));
+    // 3. Invalid hostname and port.
+    hostAndPorts.add(HostAndPort.fromParts("foo.bar", 10003));
+    // 4. Finally valid host:port
+    hostAndPorts.add(HostAndPort.fromParts(masterHostName, masterPort));
+    final String config = Joiner.on(",").join(hostAndPorts);
+    conf.set(HMasterAsyncRegistry.CONF_KEY, config);
+    conf.set(AsyncRegistryFactory.REGISTRY_IMPL_CONF_KEY, HMasterAsyncRegistry.class.getName());
+    // make sure that we do not depend on this config when getting locations for meta replicas, see
+    // HBASE-21658.
+    conf.setInt(META_REPLICAS_NUM, 1);
+    REGISTRY = new HMasterAsyncRegistry(conf);
+    customConnection = TEST_UTIL.getCustomConnection(conf);
+  }
+
+  @AfterClass
+  public static void tearDown() throws Exception {
+    IOUtils.closeQuietly(REGISTRY);
+    TEST_UTIL.shutdownMiniCluster();
+  }
+
+  @Test
+  public void testRegistryImpl() throws Exception {
+    HMaster master = TEST_UTIL.getMiniHBaseCluster().getMaster();
+    assertEquals(REGISTRY.getClusterId().get(), master.getClusterId());
+    assertEquals(REGISTRY.getMasterAddress().get(), master.getServerName());
+  }
+
+  /**
+   * Tests basic create, put, scan operations using the connection.
+   */
+  @Test
+  public void testCustomConnectionBasicOps() throws Exception {
```

Comment:
Can we not parameterize the various `TestFromClientSide*` classes to use both the new and the old methods?

Reply:
https://github.com/apache/hbase/pull/807 is a blocker to parameterize this. Will do it once that is committed.